---
# Feel free to add content and custom Front Matter to this file.
# To modify the layout, see https://jekyllrb.com/docs/themes/#overriding-theme-defaults
layout: home
title: About
nav_order: 1
has_toc: false
---

# About me

I am a plant biologist and bioinformatician interested in how plants interact with their enivornment using comparative genomic approaches. In my current role, I work as a computational and structural biologist as a SCINet ORISE Post Doctoral Fellow at the USDA in Albany, CA. My research focuses on improving protein function annotation using machine learning approaches on data obtained from Protein Language Models, protein structure predictions and deep learning generative models. Over the past year, I have developed a machine learning classifier for protein phosphorylationa and analyzed the predicted pan-interactome of 26 maize inbred lines. My past research involved studying maize immunity to pests and pathogens using molecular, biochemical, metabolomic, transcriptomic and genomic approaches. As part of that work, I have identified a candidate receptor for a herbivore-associated molecule that turns on immune responses against chewing caterpillars and charecterized a family of endogenous maize peptide hormones  that turn on immune responses. I have also contributed to work on elucidating maize antibiotic biosynthetic pathways using gene coexpression analyses and studied their role in promoting root-rhyzosphere interactions. To make my work broadly accesible, I develop Shiny and Dash data analysis web-applications, including my work on a website for plant comparative genomics, [PlantApp](https://www.plantapp.org/). I am also slowly working on adding new resources and tutorials here, so check back on occation if you are interested in any of these research topics.

# Research topics

## 1. Protein Language Models 

## PhostBoost - a machine learning classifier for protein phosphorylation using Protein Language Models

Protein phosphorylation is a post-translational modification that can alter the structure and function of proteins. PhosBoost is a machine learning approach that leverages protein language models and gradient boosting trees to predict protein phosphorylation from experimentally derived data. PhosBoost offers improved performance when recall is prioritized while consistently providing more confident probability scores.

![](assets/images/img_PhosBoost.png)

## 2. Protein-protein interactions

We deployed the PPI clustering algorithm ClusterONE to identify numerous predicted-PPI clusters that were functionally annotated using gene ontology (GO) functional enrichment, demonstrating a diverse range of enriched GO terms across different clusters. We show that the functionally annotated PPI clusters establish a useful framework for protein function prediction and prioritization of candidate genes of interest.

![](assets/images/img_PanPPI.png)

## 3. Protein structure-function analysis

Using predicted protein structures of metabolomic enzymes with generative molecular docking is a powerful tool for understanding substrate specificity and promescuity that enables the metabolomic diversity observed in plants. 

![](assets/images/img_enzyme_ligand.png)

## 4. Metabolomic profiling of plant tissues

Plant specialized metabolites play important roles in mediating beneficial interaction with microbes and insects and preventing damage from pests and disease. Extraction and profiling of plant metabolites is vital for untangling the basis for these complex ecological and agronomical interactions.

![](assets/images/img_metabolomics.png)

## 5. Association mapping to understand plant phenotypic diversity

Connecting traits to causal genes through association studies can accelarate translational research and gain insights into the underlying biological processes.

![](assets/images/img_GWAS.png)

## 6. Comparative genomic analyses across multiple genomes

Comparative genomics and phylogenemocs approaches are useful for understanding complex gene duplication, deletion and rearrangement patterns. A better understanind of these events can shed a light on gene expression regulation and neofunctionalization across the genetic diversity of plants. 

![](assets/images/img_comparative_genomics.png)