---
# Feel free to add content and custom Front Matter to this file.
# To modify the layout, see https://jekyllrb.com/docs/themes/#overriding-theme-defaults
layout: home
title: About
nav_order: 1
has_toc: false
---

# About me

Iâ€™m a biologist and bioinformatician with expertise in plant immunity, comparative genomics, and computational protein methods. During my Ph.D. I received my Ph.D. from the Huffaker lab at UC San Diego where I characterized a maize anti-herbivory receptor, peptide hormones, and metabolomic biosynthetic pathways. Currently, I am a post-doc with the USDA GrainGenes group, where I apply state-of-the-art methods in protein structure prediction, protein language models and molecular docking to improve protein function annotation through prediction of protein phosphorylation, protein-protein interactions, and protein-ligand interactions. 

# Research topics

## 1. Protein Language Models 

## PhostBoost - a machine learning classifier for protein phosphorylation using Protein Language Models

Protein phosphorylation is a post-translational modification that can alter the structure and function of proteins. PhosBoost is a machine learning approach that leverages protein language models and gradient boosting trees to predict protein phosphorylation from experimentally derived data. PhosBoost offers improved performance when recall is prioritized while consistently providing more confident probability scores.

![](assets/images/img_PhosBoost.png){: width="250" }

## 2. Protein-protein interactions

We deployed the PPI clustering algorithm ClusterONE to identify numerous predicted-PPI clusters that were functionally annotated using gene ontology (GO) functional enrichment, demonstrating a diverse range of enriched GO terms across different clusters. We show that the functionally annotated PPI clusters establish a useful framework for protein function prediction and prioritization of candidate genes of interest.

![](assets/images/img_PanPPI.png){: width="250" }

## 3. Protein structure-function analysis

Using predicted protein structures of metabolomic enzymes with generative molecular docking is a powerful tool for understanding substrate specificity and promescuity that enables the metabolomic diversity observed in plants. 

![](assets/images/img_enzyme_ligand.png){: width="250" }

## 4. Metabolomic profiling of plant tissues

Plant specialized metabolites play important roles in mediating beneficial interaction with microbes and insects and preventing damage from pests and disease. Extraction and profiling of plant metabolites is vital for untangling the basis for these complex ecological and agronomical interactions.

![](assets/images/img_metabolomics.png){: width="250" }

## 5. Association mapping to understand plant phenotypic diversity

Connecting traits to causal genes through association studies can accelarate translational research and gain insights into the underlying biological processes.

![](assets/images/img_GWAS.png)

## 6. Comparative genomic analyses across multiple genomes

Comparative genomics and phylogenemocs approaches are useful for understanding complex gene duplication, deletion and rearrangement patterns. A better understanind of these events can shed a light on gene expression regulation and neofunctionalization across the genetic diversity of plants. 

![](assets/images/img_comparative_genomics.png){: width="250" }