---
# Feel free to add content and custom Front Matter to this file.
# To modify the layout, see https://jekyllrb.com/docs/themes/#overriding-theme-defaults
layout: home
title: About
nav_order: 1
has_toc: false
---

# Plant immunity through the lens of maize comparative genetics

Maize is one of the major model systems for studying the mechanisms mediating plant immunity against pests and pathogens. Damage to crops caused by pests and pathogens is twofold: yield reduction and potential accumulation of food toxins. Studying maize immunity has been made possible, in a large part, thanks to multiple mapping populations and sequenced genomes that enable us to probe the natural genetic and phenotypic diversity across maize inbreds. One of my research interests is to build upon my experiences working on maize genetics, transcriptomics and metabolomics and expand it to other plant species, including plant species of conservation concern. I am plant biologist and bioinformatician working as a post-doc on maize biosynthesis of defensive specialized metabolites at UC San Diego with Dr. Alisa Huffaker and Dr. Eric Schmelz. In my post-doctoral research I am studying the genetic diversity of root specialized metbolites in maize and their role in protecting plants against pathogens and reshaping the root microbiome. I completed my PhD in 2021 in the Huffaker lab where I studied maize responses to elicitors that mediate anti-herbivore defense responses. For a summary of my past research I include my [CV](assets/pdfs/eporetsky_cv.pdf) and a list of my [publications](research/). The rest of my website is written as a part-blog, part-notebook, demonstrating different bioinformatics-related scripts and workflows, development of data analysis apps, and the construction of [PlantApp.org](https://www.plantapp.org), a comparative genetics database website.

# PlantApp.org: A comparative genomics website for crops, reference organisms and species of conservation concern

I have started working on PlantApp in August 2021 on my free time and have been slowly adding genomes, genomic annotations and analysis apps to it. The underlying objective of developing PlantApp was to crerate a lightweight repository that aggregates primarily gene-centric information from multiple plant genomes using a variety of information and annotation sources. To make PlantApp possible I have developed SQNce, a simple SQLite database for parsing commonly used gene-annotation files. PlantApp is openly available on [PlantApp.org](https://www.plantapp.org) and the SQNce GitHub reposity is openly available on my [SQNce GitHub page](https://github.com/eporetsky/SQNce). I am working on a comprehensive guide to PlantApp on my [PlantApp GitHub.io page](plantapp/) with additional examples available on the PlantApp page itself. For additional information, requests, suggestions and comments please get in touch with me on Twitter [@externelly](https://twitter.com/externelly) or by email at eporetsky at ucsd.edu. While I don't have a priority for genomes, annotation files or apps that I intend to add to PlantApp, I am interested in adding more of the available genomes from native plants that are used for ecological restoration projects. To see what genomes and annotation files are currently available on PlantApp visit the [available DBs page](https://www.plantapp.org/SQNce/available_dbs).