---
# Feel free to add content and custom Front Matter to this file.
# To modify the layout, see https://jekyllrb.com/docs/themes/#overriding-theme-defaults
layout: home
title: About
nav_order: 1
has_toc: false
---

# About me

Hello, I’m Elly Poretsky—a biologist and bioinformatician passionate about uncovering the molecular underpinnings of life. My expertise lies at the intersection of plant immunity, comparative genomics, and computational protein methods, where I leverage cutting-edge computational tools to address fundamental biological questions.
I earned my Ph.D. in the Huffaker Lab at UC San Diego, where I explored diverse projects, from characterizing a maize anti-herbivory receptor to investigating peptide hormones and their receptors, as well as unraveling metabolic biosynthetic pathways. This work laid the foundation for a career that bridges experimental biology with computational innovation.

Currently, I am a SCINet postdoctoral researcher with the USDA GrainGenes group In this role, I leverage cutting-edge machine learning (ML) and deep learning (DL) methods—including protein structure prediction, protein language models, and molecular docking—to improve protein function annotation and generate resources for the broader research community. My research focuses on predicting key protein attributes such as phosphorylation sites, protein–protein interactions, and protein–ligand interactions. These efforts aim to enhance our understanding of protein function and its implications for agricultural and biological research.

`For the plant researchers reading this –` please also check out the genomics database, [www.plantapp.org](www.plantapp.org), that I have been developing and maintaining. PlantApp is a canonical gene-centric database designed to facilitate comparative genomics in plants through an intuitive and user-friendly interface. Currently, PlantApp features comprehensive protein descriptions, gene families, orthogroups, domain annotations, and genomic coordinates. More recently, I have integrated a large collection of re-analyzed transcriptomic data for over 20,000 RNA-seq samples from about 15 plant species. Interactive visualizations can be used to explore gene expression patterns, differentially expressed genes and gene coexpression. PlantApp also includes a number of apps that can help you build a quick phylogeny tree, run GO enrichment analyses, and visualize domain architecture. My goal with PlantApp is to provide researchers with a centralized, accessible resource for exploring plant genomes and transcriptomes, empowering discoveries in plant biology and beyond.

For a deeper dive into my work, including my CV, publication list, and selected citations, please visit the Research tab on the left.
Thank you for stopping by!

[GitHub](https://github.com/eporetsky/){: .btn .btn-purple }
[Docker](https://hub.docker.com/u/externelly){: .btn .btn-blue }
[PlantApp](https://www.plantapp.org/){: .btn .btn-green }

[Linkedin](https://www.linkedin.com/in/eporetsky/){: .btn }
[Mastodon](https://genomic.social/@eporetsky){: .btn }

![](https://github.com/eporetsky/eporetsky.github.io/blob/master/assets/images/img_greenhouse.png?raw=true){: width="600" }